File: 345_reverse_vowels_of_a_string.java
class Solution {
    public String reverseVowels(String s) {
        if (s == null || s.length() == 0) return s;

        List<Character> vowels = new ArrayList<>(Arrays.asList('a', 'e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U'));
        StringBuilder sb = new StringBuilder(s);
        int start = 0, end = s.length() - 1;
        while(start < end) {
            char startChar = s.charAt(start);
            char endChar = s.charAt(end);
            if (vowels.contains(startChar) && vowels.contains(endChar)) {
                sb.setCharAt(start, endChar);
                sb.setCharAt(end, startChar);
                start++;
                end--;
            } else if (vowels.contains(startChar) && !vowels.contains(endChar)){
                end--;
            } else if (!vowels.contains(startChar) && vowels.contains(endChar)){
                start++;
            } else {
                start++;
                end--;
            }
        }
        return sb.toString();
    }
}

File: python/345_reverse_vowels_of_a_string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = set(["a", "e", "i", "o", "u", "A", "E", "I", "O", "U"])
        left, right = 0, len(s) - 1
        arr = list(s)
        while left < right:
            while left < right and arr[left] not in vowels: left += 1
            while left < right and arr[right] not in vowels: right -= 1
            if left < right:
                arr[left], arr[right] = arr[right], arr[left]
                left += 1
                right -= 1
        return "".join(arr)
